pkg/common/connectionstore/connection_record.go

```diff
@@ -0,0 +1,202 @@
+	rec, err := getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.store)
+	if errors.Is(err, storage.ErrDataNotFound) {
+		return getAndUnmarshal(GetConnectionKeyPrefix()(connectionID), c.transientStore)
+	}
+
+	return rec, err
+}
+
+// QueryConnectionRecords returns connection records found in underlying store
+// for given query criteria
+func (c *ConnectionLookup) QueryConnectionRecords() ([]*ConnectionRecord, error) {
+	// TODO https://github.com/hyperledger/aries-framework-go/issues/655 query criteria to be added as part of issue
+	searchKey := GetConnectionKeyPrefix()("")
+
+	itr := c.store.Iterator(searchKey, fmt.Sprintf(limitPattern, searchKey))
+	defer itr.Release()
+
+	var records []*ConnectionRecord
+
+	keys := make(map[string]struct{})
+
+	for itr.Next() {
+		var record ConnectionRecord
+
+		err := json.Unmarshal(itr.Value(), &record)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query connection records, %w", err)
+		}
+
+		keys[string(itr.Key())] = struct{}{}
+
+		records = append(records, &record)
+	}
+
+	transientItr := c.transientStore.Iterator(searchKey, fmt.Sprintf(limitPattern, searchKey))
+	defer transientItr.Release()
+
+	for transientItr.Next() {
+		// don't fetch data from transient store if same record is present in permanent store
+		if _, ok := keys[string(transientItr.Key())]; ok {
+			continue
+		}
+
+		var record ConnectionRecord
+
+		if err := json.Unmarshal(transientItr.Value(), &record); err != nil {
+			return nil, fmt.Errorf("query connection records from transient store : %w", err)
+		}
+
+		records = append(records, &record)
+	}
+
+	return records, nil
+}
+
+// GetConnectionRecordAtState return connection record based on the connection ID and state.
+func (c *ConnectionLookup) GetConnectionRecordAtState(connectionID, stateID string) (*ConnectionRecord, error) {
+	if stateID == "" {
+		return nil, errors.New(stateIDEmptyErr)
+	}
+
+	return getAndUnmarshal(GetConnectionStateKeyPrefix()(connectionID, stateID), c.transientStore)
+}
+
+func getAndUnmarshal(k string, store storage.Store) (*ConnectionRecord, error) {
+	connRecordBytes, err := store.Get(k)
+	if err != nil {
+		return nil, err
+	}
+
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// GetConnectionRecordByNSThreadID return connection record via namespaced threadID
+func (c *ConnectionLookup) GetConnectionRecordByNSThreadID(nsThreadID string) (*ConnectionRecord, error) {
+	connectionIDBytes, err := c.transientStore.Get(nsThreadID)
+	if err != nil {
+		return nil, fmt.Errorf("get connectionID by namespaced threadID: %w", err)
+	}
+	// adding prefix for storing connection record
+	k := GetConnectionKeyPrefix()(string(connectionIDBytes))
+
+	connRecordBytes, err := c.transientStore.Get(k)
+	if err != nil {
+		return nil, fmt.Errorf("get connection record by connectionID: %w", err)
+	}
+
+	return prepareConnectionRecord(connRecordBytes)
+}
+
+// Store returns handle to underlying permanent store
```

Comment:
Are these used?

Reply:
Yes, private `connectionStore` inside didexchange service  which wraps this struct calls Put() & Get() on these store instances ( Store() & TransientStore() functions).
These two `Store() and TransientStore()` will be removed once we move those custom save/update logic out of didexchange service package.